### ISSUE TITLE
[Enhancement]: IBM Power (ppc64le) support

### ISSUE THREAD

**Issue by sumitd2** [milvus-io/milvus#29566]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

I belong to the IBM Power porting team. We have recently developed patches to build Milvus on ppc64le, and would like to open a PR to upstream the changes to this repo. The major changes were adjusting the versions for some conan installed dependencies, and creating a local cmake repo for conan because the conan binary package for cmake currently does not have power binaries. So we need to somehow find a way to ensure cmake supports power, and any future dependency version modifications work on all architectures. Do you have any suggestions?

### Why is this needed?

There are customers who wish to use Milvus on Power.

### Anything else?

https://github.com/ppc64le/build-scripts/pull/3467

**Comment by sumitd2** [comment]
cc: @seth-priya

**Comment by xiaofan-luan** [comment]
> ### Is there an existing issue for this?
> * [x]  I have searched the existing issues
> 
> ### What would you like to be added?
> I belong to the IBM Power porting team. We have recently developed patches to build Milvus on ppc64le, and would like to open a PR to upstream the changes to this repo. The major changes were adjusting the versions for some conan installed dependencies, and creating a local cmake repo for conan because the conan binary package for cmake currently does not have power binaries. So we need to somehow find a way to ensure cmake supports power, and any future dependency version modifications work on all architectures. Do you have any suggestions?
> 
> ### Why is this needed?
> There are customers who wish to use Milvus on Power.
> 
> ### Anything else?
> [ppc64le/build-scripts#3467](https://github.com/ppc64le/build-scripts/pull/3467)

Great work.
The only concern is how we can deploy ci/cid to make sure the Power deployment will not be broken by some new PRs.
We don't have any IBM Power machines so that might be really challenge

**Comment by sumitd2** [comment]
Hi @xiaofan-luan You can a ppc64le node for your CI from this link:
https://osuosl.org/services/powerdev/request_hosting/    (Please put "Gerrit Huizenga" in the IBM advocate field.)

Also, I have successfully built and tested Milvus 2.3.3 using our patches on Ubuntu 22.04 and RHEL 9.3, on both x86_64 and ppc64le. You will need to create a repo for "cmake/3.28.3@milvus/dev" repo to host cmake binaries for both Intel and Power. I hope that's fine.

cc: @seth-priya

**Comment by xiaofan-luan** [comment]
keep this

**Comment by xiaofan-luan** [comment]
@sumitd2 
sorry for the late reply.

I'm not very familiar with the compilation part. We can keep it as long as there is simple script that can run for milvus to build on Power series. 

This seems to be a large pr so let's try to do this
1. change knowhere to support POWER -> especially for the same functionality of SIMD
2. make a script that can compile milvus 
3. dockerfile

**Comment by sumitd2** [comment]
Hi @xiaofan-luan, we are working on the SIMD optimizations and the PR. Will get back to you on this.

**Comment by xiaofan-luan** [comment]
@alexanderguzhva 
can help on this if necessary

**Comment by alexanderguzhva** [comment]
@sumitd2 I've noticed a wave of AIX-related PRs to the Faiss baseline. I will propagate these changes to our Faiss fork in Knowhere. Please feel free to let me know if you need any assistance with Faiss/Knowhere as well.

**Comment by sumitd2** [comment]
> @sumitd2 sorry for the late reply.
> 
> I'm not very familiar with the compilation part. We can keep it as long as there is simple script that can run for milvus to build on Power series.
> 
> This seems to be a large pr so let's try to do this
> 
> 1. change knowhere to support POWER -> especially for the same functionality of SIMD
> 2. make a script that can compile milvus
> 3. dockerfile

Hi @xiaofan-luan, couple of quick points here:

1. Our team that was working on SIMD has abandoned the idea. They say the compiler generated code is better than handwritten.
2. A script here will not work, as it has the chance of being broken by future PRs. This will have to be a proper port with the changes going into master, and a CI node (which we can provide on osuosl.org) to include the Power CI as a job. The conan binary package for cmake currently does not have Power binaries, so we will have to find a solution for this - many of our customers are interested in Milvus on Power

Let me know your thoughts. cc: @seth-priya

**Comment by xiaofan-luan** [comment]
@alexanderguzhva could you help with the IBM team with the milvus side change?

**Comment by xiaofan-luan** [comment]
@locustbaby could you try to request a powerPC machine for the basic ci/cd?

**Comment by xiaofan-luan** [comment]
> > @sumitd2这么晚才回复很抱歉。
> > 我对编译部分不是很熟悉。只要有简单的脚本可以运行 milvus 在 Power 系列上构建，我们就可以保留它。
> > 这似乎是一个很大的公关，所以让我们尝试这样做
> > 
> > 1. 更改knowhere以支持POWER -> 特别是对于SIMD的相同功能
> > 2. 制作一个可以编译milvus的脚本
> > 3. docker文件
> 
> 你好@xiaofan-luan，这里有几个要点：
> 
> 1. 我们致力于 SIMD 的团队已经放弃了这个想法。他们说编译器生成的代码比手写的更好。
> 2. 这里的脚本不起作用，因为它有可能被未来的 PR 破坏。这必须是一个正确的端口，更改将进入主节点，并且必须是一个 CI 节点（我们可以在 osuosl.org 上提供）以将 Power CI 作为作业包含在内。 cmake 的 conan 二进制包目前没有 Power 二进制文件，因此我们必须为此找到解决方案 - 我们的许多客户对 Milvus on Power 感兴趣
> 
> 让我知道你的想法。抄送：@seth-priya



> > @sumitd2 sorry for the late reply.
> > I'm not very familiar with the compilation part. We can keep it as long as there is simple script that can run for milvus to build on Power series.
> > This seems to be a large pr so let's try to do this
> > 
> > 1. change knowhere to support POWER -> especially for the same functionality of SIMD
> > 2. make a script that can compile milvus
> > 3. dockerfile
> 
> Hi @xiaofan-luan, couple of quick points here:
> 
> 1. Our team that was working on SIMD has abandoned the idea. They say the compiler generated code is better than handwritten.
> 2. A script here will not work, as it has the chance of being broken by future PRs. This will have to be a proper port with the changes going into master, and a CI node (which we can provide on osuosl.org) to include the Power CI as a job. The conan binary package for cmake currently does not have Power binaries, so we will have to find a solution for this - many of our customers are interested in Milvus on Power
> 
> Let me know your thoughts. cc: @seth-priya

we also need your help to finish the power pc compile. is there a pr for this issue?

**Comment by xiaofan-luan** [comment]
sure @locustbaby can help you to do the setup

**Comment by locustbaby** [comment]
@sumitd2 I've submitted a ticket#33429 for a powerpc

**Comment by sumitd2** [comment]
> @sumitd2 I've submitted a ticket#33429 for a powerpc

@locustbaby We have received your request, and it has been approved. Now its a matter of the administrator creating the VM, shouldn't be long.

**Comment by sumitd2** [comment]
@locustbaby OSL has created the VM but they are waiting on a GPG key from you to provide access:
> Your Openstack account is set up as well. Please send us your public GPG key,
> in order to allow us to securely provide you the login details.
> Checking back to see if you were able to log in and send us a gpg key.

**Comment by locustbaby** [comment]
@sumitd2 I'm not very familiar with OSL, where should I submit the GPG key, I didn't receive any email about this.
And should I add the GPG key to our project? It would be great if you could describe the steps to follow!

**Comment by janani66** [comment]
@locustbaby -- OSU admins need a public gpg keys from you so that we can send the VPN certificates for the VM in an encrypted message. Please respond in the [ticket](https://support.osuosl.org/Ticket/Display.html?id=33429) you had created with your GPG key.

**Comment by locustbaby** [comment]
@janani66 How to login the Support channel, where can I register it？I didn't have an account.

**Comment by janani66** [comment]
How did you create the ticket without login creds?

Anyway you can also paste your public gpg key here in this issue and I can pass it on to the OSU folks

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Sheldon ***@***.***>
Sent: Friday, May 17, 2024 1:32:58 AM
To: milvus-io/milvus ***@***.***>
Cc: Janani Janakiraman ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [milvus-io/milvus] [Enhancement]: IBM Power (ppc64le) support (Issue #29566)

@janani66 How to login the Support channel, where can I register it？I didn't have an account. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned. Message ID: <milvus-io/milvus/issues/29566/2116849448@ github. com>
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.
<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/AdhS1Rd-!91FQ2X26mqcYXt6aZU-rGealfi8_DtOaYyeYvgQJzoalVbvypqqcq6jypjhqEnvVBYxnSk5TDOhaIiK_U0k4sQbz1jIZ9-ur0ykDZQ$>
Report Suspicious

ZjQcmQRYFpfptBannerEnd

@janani66<https://github.com/janani66> How to login the Support channel, where can I register it？I didn't have an account.

**Comment by locustbaby** [comment]
@janani66 No login required for the ticket, I just submitted it directly.
Here is my GPG key, please guide me continue, thanks
[gpg.txt](https://github.com/milvus-io/milvus/files/15349958/gpg.txt)

**Comment by alexanderguzhva** [comment]
Please make sure that the corresponding PowerPC patches for knowhere are present in your codebase during your experiments:
* https://github.com/zilliztech/knowhere/pull/562
* https://github.com/zilliztech/knowhere/pull/524
* https://github.com/zilliztech/knowhere/pull/526  

@locustbaby @sumitd2

**Comment by zwall-bp** [comment]
Hello @locustbaby I'm Zach, with the OSL. Just wanted to extend out to let you know I have sent your credentials to your email. Be sure to check for any messages from `powerdev-request@osuosl.org` in the inbox of the account your provided us when your sent the request. It may have gone into your spam folder.

**Comment by locustbaby** [comment]
@zwall-bp Thanks for all the messages in the email, Zach, I can connect the VM by ssh now,
I used to receive an email that told me the ticket was created, but as I said, I submitted the request without any login so that I never received messages before.
![image](https://github.com/milvus-io/milvus/assets/21237232/8221d2b0-391e-4f13-a985-c2b1d662644c)

**Comment by locustbaby** [comment]
@sumitd2 What's the next step I need to do? Add this VM to this project?
@zwall-bp It would be very nice if you could guide me registration, I retried https://osuosl.org/services/powerdev/request_hosting/ again and found out that even an empty form can be submitted.

**Comment by zwall-bp** [comment]
I have already set up a VM for you to use. I'm assuming by registration, you are asking about getting access to our OpenStack Web UI, which allows you to remove your currently running VM, and create a fresh, new one. [More information about utilizing OpenStack can be found on our wiki](http://wiki.osuosl.org/openstack/creating_instances.html). My latest email has the encrypted message for logging into OpenStack. [More information about decrypting the file can be found here](https://www.gnupg.org/gph/en/manual/x110.html).

If the registration is about our support site, we only use that internally, and communicate with you through email.

**Comment by sumitd2** [comment]
@locustbaby I have a script with a patch that builds milvus (2.4.1) and runs the go, cpp and e2e tests. There are a few failing tests, so we will have to figure them out, 10 Azure related cpp tests are failing, 3 golang tests, and 3 e2e tests (xpassing). Let me know when you are ready.
cc: @seth-priya

**Comment by xiaofan-luan** [comment]
> @locustbaby I have a rhel script with a patch that builds milvus (2.4.1) and runs the go, cpp and e2e tests. There are a few failing tests, so we will have to figure them out, 10 Azure related cpp tests are failing, 3 golang tests, and 3 e2e tests (xpassing). Let me know when you are ready. cc: @seth-priya @alexanderguzhva

we can ignore azure test for now. becasue i guess no one is actually running on azure + ppc? if you can share all the failing case we can help on check it as well

**Comment by sumitd2** [comment]
@xiaofan-luan 3 E2e tests are xpassing:
```
milvus_client/test_milvus_client_collection.py::TestMilvusClientCollectionValid.test_milvus_client_collection_fast_creation_default[32768] X0% ▏         ilvus_client/test_milvus_client_collection.py
milvus_client/test_milvus_client_collection.py::TestMilvusClientCollectionValid.test_milvus_client_collection_fast_creation_default[128] X1% ▎          milvus_client/test_milvus_client_collection.py
milvus_client/test_milvus_client_collection.py::TestMilvusClientCollectionValid.test_milvus_client_collection_fast_creation_default[2] X1% ▏         SS milvus_client/test_milvus_client_collection.py

```
I also built it on an Intel VM, the result was the same.

**Comment by sumitd2** [comment]
@xiaofan-luan The 10 Azure related tests (same result on x86, some Azure tests are passing as well):
[  FAILED  ] AzureChunkManagerTest.WrongConfig
[  FAILED  ] AzureChunkManagerTest.AzureLogger
[  FAILED  ] AzureChunkManagerTest.BasicFunctions
[  FAILED  ] AzureChunkManagerTest.BucketPositive
[  FAILED  ] AzureChunkManagerTest.BucketNegtive
[  FAILED  ] AzureChunkManagerTest.ObjectExist
[  FAILED  ] AzureChunkManagerTest.WritePositive
[  FAILED  ] AzureChunkManagerTest.ReadPositive
[  FAILED  ] AzureChunkManagerTest.RemovePositive
[  FAILED  ] AzureChunkManagerTest.ListWithPrefixPositive

**Comment by xiaofan-luan** [comment]
I think we can xpass azure related test for now

**Comment by sumitd2** [comment]
@xiaofan-luan Failing Golang tests (I started only minio and etcd pods, rest images aren't available):

```
FAIL    github.com/milvus-io/milvus/internal/datanode   16.504s
FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/kafka    58.504s
FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/pulsar   54.918s
FAIL    github.com/milvus-io/milvus/pkg/util/etcd       0.184s
FAIL    github.com/milvus-io/milvus/cmd/tools/config    0.615s
```

**Comment by sumitd2** [comment]
Hi @xiaofan-luan @alexanderguzhva @locustbaby How would you like to proceed? I hope you now have access to the Power VM.

**Comment by locustbaby** [comment]
@sumitd2 Yeah, I can acess the VM by both SSH/Openstack

**Comment by sumitd2** [comment]
@xiaofan-luan I have observed the go tests are flaky. What can we do about the failing ones?

**Comment by xiaofan-luan** [comment]
@sumitd2

**Comment by xiaofan-luan** [comment]
can you offer the ci/cd logs so we can take a look?

Some of those are dependency related test and you need to bring up kafka/pulsar or etcd

**Comment by xiaofan-luan** [comment]
any way, please offer logs and we can take. a look.

**Comment by sumitd2** [comment]
@xiaofan-luan Here is the output of make test-go :

[milvus-v2.4.1-test-go.log](https://github.com/user-attachments/files/15758512/milvus-v2.4.1-test-go.log)
[docker-compose.yam.txt](https://github.com/user-attachments/files/15758527/docker-compose.yam.txt)

**Comment by sumitd2** [comment]
> @xiaofan-luan Failing Golang tests (I started only minio and etcd containers, rest images aren't available):
> 
> ```
> FAIL    github.com/milvus-io/milvus/internal/datanode   16.504s
> FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/kafka    58.504s
> FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/pulsar   54.918s
> FAIL    github.com/milvus-io/milvus/pkg/util/etcd       0.184s
> FAIL    github.com/milvus-io/milvus/cmd/tools/config    0.615s
> ```

@xiaofan-luan I have analyzed the failed ones:
1. github.com/milvus-io/milvus/internal/datanode   16.504s
This is flaky. I have seen it pass atleast once.

2. github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/kafka    58.504s
3. github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/pulsar   54.918s
The fail because their images are not available.

4. github.com/milvus-io/milvus/pkg/util/etcd       0.184s
If I kill the etcd container, this one passes.

5. github.com/milvus-io/milvus/cmd/tools/config    0.615s
Looks like this one fails because because configs/milvus.yaml is not updated according to paramtable.

**Comment by xiaofan-luan** [comment]
@alexanderguzhva could you help on it?

**Comment by xiaofan-luan** [comment]
@locustbaby could you help verify all ut issue?

**Comment by alexanderguzhva** [comment]
@sumitd2 Would you be able to provide the build logs for the master branch, if possible?

**Comment by sumitd2** [comment]
Hi @alexanderguzhva The logs are attached.
```
[make.log](https://github.com/user-attachments/files/15967563/make.log)
```

**Comment by alexanderguzhva** [comment]
@sumitd2 my bad. 
I know what the problem is.

**Comment by alexanderguzhva** [comment]
I don't access to test machine, but I think that https://github.com/zilliztech/knowhere/pull/665 should fix the problem

**Comment by sumitd2** [comment]
@alexanderguzhva FYI Now its building fine

**Comment by sumitd2** [comment]
@locustbaby @xiaofan-luan 
I ran the build and tests on the latest master today and got the following failures. Can you please help? 
```
FAIL    github.com/milvus-io/milvus/internal/datanode
FAIL    github.com/milvus-io/milvus/internal/storage
FAIL    github.com/milvus-io/milvus/pkg/mq/msgdispatcher
FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream
FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/pulsar
FAIL    github.com/milvus-io/milvus/pkg/streaming/walimpls/impls/pulsar
FAIL    github.com/milvus-io/milvus/cmd/tools/config
```

**Comment by xiaofan-luan** [comment]
> @locustbaby @xiaofan-luan I ran the build and tests on the latest master today and got the following failures. Can you please help?
> 
> ```
> FAIL    github.com/milvus-io/milvus/internal/datanode
> FAIL    github.com/milvus-io/milvus/internal/storage
> FAIL    github.com/milvus-io/milvus/pkg/mq/msgdispatcher
> FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream
> FAIL    github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/pulsar
> FAIL    github.com/milvus-io/milvus/pkg/streaming/walimpls/impls/pulsar
> FAIL    github.com/milvus-io/milvus/cmd/tools/config
> ```

Is there a way we can see the ci/cd and logs?

From the case it seems that Power CPU didn't support to deploy pulsar.

**Comment by sumitd2** [comment]
> @xiaofan-luan Here is the output of `make test-go` :
> 
> [milvus-v2.4.1-test-go.log](https://github.com/user-attachments/files/15758512/milvus-v2.4.1-test-go.log) [docker-compose.yam.txt](https://github.com/user-attachments/files/15758527/docker-compose.yam.txt)

@xiaofan-luan There is no ci, I run it manually on a power vm. You can refer to the last log file as above. Will have to rerun the build for the latest one.

**Comment by sumitd2** [comment]
@xiaofan-luan 

I got these four passing by using a third part pulsar image for Power:

```
ok    github.com/milvus-io/milvus/pkg/mq/msgdispatcher
ok    github.com/milvus-io/milvus/pkg/mq/msgstream
ok    github.com/milvus-io/milvus/pkg/mq/msgstream/mqwrapper/pulsar
ok    github.com/milvus-io/milvus/pkg/streaming/walimpls/impls/pulsar
```

So now there are 3 left:

1. `> FAIL    github.com/milvus-io/milvus/internal/datanode`
This one is printing:
```
--- FAIL: TestDataNode (0.01s)
panic: fail to init rocksmq [recovered]
        panic: fail to init rocksmq

```

2. `> FAIL    github.com/milvus-io/milvus/internal/storage`
This one looks Azure related. Is there any additional setup required to run this? Can you confirm?

3. `> FAIL    github.com/milvus-io/milvus/cmd/tools/config`
I think you will have to fix this. Can you confirm? Its printing:
```
--- FAIL: TestYamlFile (0.26s)
    generate_test.go:44:
                Error Trace:    /sumit/milvus/cmd/tools/config/generate_test.go:44
                Error:          configs/milvus.yaml is not consistent with paramtable, file: [autoindex:], code: [# Licensed to the LF AI & Data foundation under one]. Do not edit milvus.yaml directly.
                Test:           TestYamlFile

```

Thanks.

**Comment by xiaofan-luan** [comment]
1. we need log for this run to understand why rocksdb can not work
2.ignore azure for now is ok, do we have log for that?
3. do we have pr already? I need to understand what part of this yaml has been changed.

Please generate a pr and I will start review on this

**Comment by sumitd2** [comment]
https://github.com/milvus-io/milvus/pull/34978

**Comment by sumitd2** [comment]
@alexanderguzhva This line is causing a missing header error during the power build:
https://github.com/zilliztech/knowhere/blob/38e4c88ce2ffda1d6ca0bfed0afe3a34743ce0c5/include/knowhere/comp/thread_pool.h#L16
If I install openblas-devel and change it to #include <**openblas/**cblas.h> it builds fine.

**Comment by alexanderguzhva** [comment]
@sumitd2 will issue a fix shortly

**Comment by alexanderguzhva** [comment]
@sumitd2 if the following `#ifdef` sufficient? Or would you like me to use a particular one,?
```
#if defined(__PPC64__) || defined(__ppc64__) || defined(__PPC64LE__) || defined(__ppc64le__)
```
Thanks.

**Comment by sumitd2** [comment]
also add __powerpc64__

**Comment by sumitd2** [comment]
@alexanderguzhva it worked, thank you!

There is another error during the build (`make`). Not sure if you can fix it but posting it:

```
Building Milvus ...
# github.com/bytedance/sonic/internal/rt
/root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:102:6: missing function body
/root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:104:6: missing function body
/root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:114:6: missing function body
make: *** [Makefile:82: milvus] Error 1

```

**Comment by alexanderguzhva** [comment]
@sumitd2 yep, I believe that this is out of my immediate knowledge, sorry about that

**Comment by xiaofan-luan** [comment]
> @alexanderguzhva it worked, thank you!
> 
> There is another error during the build (`make`). Not sure if you can fix it but posting it:
> 
> ```
> Building Milvus ...
> # github.com/bytedance/sonic/internal/rt
> /root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:102:6: missing function body
> /root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:104:6: missing function body
> /root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:114:6: missing function body
> make: *** [Makefile:82: milvus] Error 1
> ```
> 
> @xiaofan-luan

@haorenfsa see that there is a problem for sonic on ppc arch. guess it dosn't support ppc as well. can we compile it as an optional?

**Comment by sumitd2** [comment]
> > @alexanderguzhva it worked, thank you!
> > There is another error during the build (`make`). Not sure if you can fix it but posting it:
> > ```
> > Building Milvus ...
> > # github.com/bytedance/sonic/internal/rt
> > /root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:102:6: missing function body
> > /root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:104:6: missing function body
> > /root/go/pkg/mod/github.com/bytedance/sonic@v1.9.1/internal/rt/gcwb.go:114:6: missing function body
> > make: *** [Makefile:82: milvus] Error 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @xiaofan-luan
> 
> @haorenfsa see that there is a problem for sonic on ppc arch. guess it dosn't support ppc as well. can we compile it as an optional?

@xiaofan-luan I am currently working on building master, and have already fixed this error by creating a sonic_ppc64le.go which uses encoding/json.

**Comment by haorenfsa** [comment]
@sumitd2 to save the work, I think we could just disable sonic json for ppc64le.

**Comment by haorenfsa** [comment]
@sumitd2 I just checked the code, and it's already optional. So we can just use `make MILVUS_GO_BUILD_TAGS=dynamic` to disable sonic.
<img width="751" alt="image" src="https://github.com/user-attachments/assets/d6bbc283-b2d3-4aa9-9511-90af62602ce2">

**Comment by sumitd2** [comment]
> @sumitd2 I just checked the code, and it's already optional. So we can just use `make MILVUS_GO_BUILD_TAGS=dynamic` to disable sonic. 

@haorenfsa That will disable it for every architecture, so why would you want to do that? I have already done the work, just needs a PR.

**Comment by haorenfsa** [comment]
> > @sumitd2 I just checked the code, and it's already optional. So we can just use `make MILVUS_GO_BUILD_TAGS=dynamic` to disable sonic.
> 
> @haorenfsa That will disable it for every architecture, so why would you want to do that? I have already done the work, just needs a PR.

Um, I think you may have misunderstood. What I meant is that you can only turn it off when compiling on powerPC. We still turn it on by default for other architectures.

**Comment by haorenfsa** [comment]
And of course it's great that you implement it for ppc.  Your patch is very welcomed.

But for other not covered architectures, like `mips` or `riscv` maybe, if one also want to build milvus on that, one can use the command I mentioned above. There won't be any difference as long as one use GRPC sdk not RESTful API.

**Comment by sumitd2** [comment]
@xiaofan-luan 
I have the latest master building on Power, and the C++ tests are passing (except for the Azure ones). This feels like the right time to set up your CI. Unfortunately GHA runner cannot be installed on the OSU VM we have provided you because it does not support Power. So the only option I see is Buildkite, which is free for open source projects. It integrates well with GHA so you will be able to see the Power job in the list on the PR page. Let me know if you are fine with it.

@locustbaby 
I need access to the OSU VM that has been assigned to you, to be able to setup the CI agent for the Power builds.

cc: @seth-priya

**Comment by xiaofan-luan** [comment]
@chuanfeng

could you helo to invesigate on that? Let's merge this pr asap

**Comment by xiaofan-luan** [comment]
@sumitd2 @locustbaby 
what's the progress of this issue?

**Comment by sumitd2** [comment]
@locustbaby I need access to the ppc64le OSU server you were assigned. Thanks.

**Comment by locustbaby** [comment]
Sorry missed, do you need the key?在 2024年12月2日，23:09，sumitd2 ***@***.***> 写道：﻿
@locustbaby I need access to the ppc64le OSU server you were assigned. Thanks.

**Comment by sumitd2** [comment]
yes, I need the IP and the credentials. You can send it on sumit.dubey2 at ibm dot com

**Comment by locustbaby** [comment]
@sumitd2 Just sent it to you, contact me if any issue.

**Comment by sumitd2** [comment]
Hi @alexanderguzhva , I am in the process of setting up a ci job for ppc64le which would run on an ubuntu 22.04 agent, and getting the following errors while building knowhere:

1. <openblas/cblas.h> not found
https://github.com/zilliztech/knowhere/pull/754 
Just <cblas.h> is needed on ubuntu. <openblas/cblas.h> is required on RHEL.

2. 
xmmintrin.h
#error "Please read comment above.  Use -DNO_WARN_X86_INTRINSICS to disable this error."
```
This header is distributed to simplify porting x86_64 code that
   makes explicit use of Intel intrinsics to powerpc64le.
   It is the user's responsibility to determine if the results are
   acceptable and make additional changes as necessary.
   Note that much code that uses Intel intrinsics can be rewritten in
   standard C or GNU C extensions, which are more portable and better
   optimized across multiple targets.
   ...
```
This will have to be defined to silence this error.

Can you please help fix them? Thanks.

**Comment by alexanderguzhva** [comment]
@sumitd2 I got a bit confused. 
1. Do I get it correct that the include file must take a linux distro into account? Some construct that has the following result:
```C++
#ifdef USE_RPM
#include <openblas/cblas.h>
#elif USE_DEB
#include <cblas.h>
#else
static_assert("?!!");
#endif
``` 
2. It is needed to add `NO_WARN_X86_INTRINSICS` to cmake flags for powerpc, right?
 
Thanks

**Comment by sumitd2** [comment]
> [@sumitd2](https://github.com/sumitd2) I got a bit confused.
> 
> 1. Do I get it correct that the include file must take a linux distro into account? Some construct that has the following result:
> 
> #ifdef USE_RPM
> #include <openblas/cblas.h>
> #elif USE_DEB
> #include <cblas.h>
> #else
> static_assert("?!!");
> #endif
> 2. It is needed to add `NO_WARN_X86_INTRINSICS` to cmake flags for powerpc, right?
> 
> Thanks

[alexanderguzhva](https://github.com/alexanderguzhva) Sorry I forgot to respond. I eventually did it on RHEL. Thanks anyway.
